src/qtism/common/datatypes/QtiCoords.php

```diff
@@ -120,52 +120,58 @@ public function inside(QtiPoint $point)
     {
         if ($this->getShape() === QtiShape::DEF) {
             return true;
-        } elseif ($this->getShape() === QtiShape::RECT) {
+        }
+
+        if ($this->getShape() === QtiShape::RECT) {
             return $point->getX() >= $this[0] && $point->getX() <= $this[2] && $point->getY() >= $this[1] && $point->getY() <= $this[3];
-        } elseif ($this->getShape() === QtiShape::CIRCLE) {
+        }
+
+        if ($this->getShape() === QtiShape::CIRCLE) {
             return pow($point->getX() - $this[0], 2) + pow($point->getY() - $this[1], 2) < pow($this[2], 2);
-        } else {
-            // we consider it is a polygon.
-            // - Transform coordinates in vertices.
-            // -- Use of the "point in polygon" algorithm.
-            $vertices = [];
-            for ($i = 0; $i < count($this); $i++) {
-                $vertex = [];
-                $vertex[] = $this[$i]; //x
-                $i++;
-                $vertex[] = $this[$i]; //y
-
-                $vertices[] = $vertex;
+        }
+
+        // we consider it is a polygon.
+        // - Transform coordinates in vertices.
+        // -- Use of the "point in polygon" algorithm.
+        $vertices = [];
+        for ($i = 0, $limit = $this->count(); $i < $limit; $i++) {
+            $vertex = [];
+            $vertex[] = $this[$i]; //x
+            $i++;
+            $vertex[] = $this[$i]; //y
+
+            $vertices[] = $vertex;
+        }
+
+        $intersects = 0;
+        for ($limit = count($vertices), $i = 0, $j = $limit - 1; $i < $limit; $i++) {
```

Comment:
Nice trick to make it loop between first and last vertices. ;-)
Please just move the additional initialisations out of the for statement, it's difficult to read:
```suggestion
        $limit = count($vertices);
        $j = $limit - 1;
        for ($i = 0; $i < $limit; $i++) {
```